tests/test_server.py

```diff
@@ -52,140 +71,152 @@ def test_add_device(client, name='steelseries-rival310'):
-    assert 'name' in data
+    client.delete('/devices/{}'.format(data['id']))
 
 
 def test_delete_device(client):
-    id = test_add_device(client)
+    id = add_device(client)
 
     response = client.delete('/devices/{}'.format(id))
     assert response.status_code == 204
 
 
-def test_get_dpi(client, dpi_id=0):
-    from ratbag_emu.device_handler import DeviceHandler
-
-    id = test_add_device(client)
+def test_get_device(client):
+    id = add_device(client)
 
-    response = client.get('/devices/{}/phys_props/dpi/{}'.format(id, dpi_id))
+    response = client.get('/devices/{}'.format(id))
     assert response.status_code == 200
-    data = json.loads(response.data)
-    assert data == DeviceHandler.devices[id].hprof.dpi[dpi_id if
-        dpi_id != 'active' else DeviceHandler.devices[id].hprof.active_dpi]
+    data = response.json()
+    assert 'id' in data
+    assert 'shortname' in data
+    assert 'name' in data
+    assert 'input_nodes' in data
 
+    client.delete('/devices/{}'.format(id))
 
-def test_set_dpi(client, dpi_id=0):
+
+def test_dpi(client, dpi_id=0):
     from ratbag_emu.device_handler import DeviceHandler
 
-    id = test_add_device(client)
+    id = add_device(client)
 
-    DeviceHandler.devices[id].hprof.dpi[dpi_id if dpi_id != 'active' else
-        DeviceHandler.devices[id].hprof.active_dpi] = 6666
+    data = 6666
+    response = client.put('/devices/{}/phys_props/dpi/{}'.format(id, dpi_id),
+                          json=data)
 
     response = client.get('/devices/{}/phys_props/dpi/{}'.format(id, dpi_id))
     assert response.status_code == 200
-    data = json.loads(response.data)
-    assert data == 6666
-
+    assert response.json() == data
 
-def test_get_active_dpi(client):
-    test_get_dpi(client, 'active')
+    client.delete('/devices/{}'.format(id))
 
 
-def test_set_active_dpi(client):
-    test_set_dpi(client, 'active')
+def test_active_dpi(client):
+    test_dpi(client, 'active')
 
 
-def test_get_led(client):
-    id = test_add_device(client)
-
-    response = client.get('/devices/{}/phys_props/leds/0'.format(id))
-    assert response.status_code == 200
-
-
-def test_set_led(client):
-    id = test_add_device(client)
+def test_led(client):
+    id = add_device(client)
 
     data = [
         0xFF,
         0xFF,
         0xFF
     ]
-    response = client.put('/devices/{}/phys_props/leds/0'.format(id),
-                          data=json.dumps(data),
-                          content_type='application/json')
+    response = client.put('/devices/{}/phys_props/leds/0'.format(id), json=data)
+    assert response.status_code == 200
+
+    response = client.get('/devices/{}/phys_props/leds/0'.format(id))
     assert response.status_code == 200
+    assert response.json() == data
+
+    client.delete('/devices/{}'.format(id))
 
 
-def test_device_event_raw_xy(client):
+def test_device_event(client):
     import os
 
-    from time import sleep
+    from time import time, sleep
```

Comment:
fwiw, localised imports don't really give you that much benefit unless you don't expect that part of the code to be called. Doubly so for full module  imports like `import libevdev` - you'll have to use the module name anyway so just import them at the top of the file.

And in many cases not using `import from` means the namespace provides extra benefit to the reader because it's clearer where a class comes from.